### ISSUE TITLE
install_multiple_protocol_interfaces definition is NOT variadic

### ISSUE THREAD

**Issue by jyao1** [r-util/r-efi#12]
The UEFI specification defines install_multiple_protocol_interfaces as variadic.
The definition below shows it does not follow UEFI specification.

May I know if there is any plan to add such support it?

    pub install_multiple_protocol_interfaces: eficall!{fn(
        *mut crate::base::Handle,
        // XXX: Actual definition is variadic. See eficall!{} for details.
        *mut core::ffi::c_void,
        *mut core::ffi::c_void,
    ) -> crate::base::Status},
    pub uninstall_multiple_protocol_interfaces: eficall!{fn(
        *mut crate::base::Handle,
        // XXX: Actual definition is variadic. See eficall!{} for details.
        *mut core::ffi::c_void,
        *mut core::ffi::c_void,
    ) -> crate::base::Status},

**Comment by dvdhrm** [comment]
Hey!

I tried to explain this in the `eficall` macro as mentioned in the comment directly above the definition:

    https://docs.rs/r-efi/2.1.0/r_efi/macro.eficall.html#variadics

The rust compiler does not support this so far. Is there a strong reason why you need these calls? There are alternatives to these functions which should work fine.

Thanks
David

**Comment by jyao1** [comment]
Thanks.
I ask this because I am trying to write a UEFI implementation to produce these services.
I think r_efi is good place to start, because it already defines some UEFI interface.

Today, the gBS->InstallMultipleProtocolInterfaces() are widely used in UEFI drivers.
In EDKII project, https://github.com/tianocore/edk2/tree/master/MdeModulePkg has 112  InstallMultipleProtocolInterfaces() function calls, including the PCI, USB, ATA, NVMe, console, variable modules etc.

Of course, we can use InstallProtocolInterface() one by one to install all those required protocols. However, the error handling will become very complicated. For example, if 1st and 2nd protocols are installed successfully but the 3rd protocol fails, then we have to uninstall the 1st and 2nd protocols. It adds burden for developer, and that is why IntallMultipleProtocolInterfaces() is used.

Currently, the maximal number of parameter used the of IntellMultipleProtocolInterfaces() in EDKII is 6 GUID/Interface pairs, which means 14 parameters (plus first handle, and last NULL terminator)

If we can have a straight forward way to achieve this, it will be great.

Thank you
Yao Jiewen

**Comment by dvdhrm** [comment]
Oh, yeah, I understand that the function is very convenient and I would very much like to support it. My point simply was that it is not strictly necessary. The EDK2 implementation of that function is less than 100 lines (see `CoreInstallMultipleProtocolInterfaces()`), so it would be rather straightforward to provide it in a helper crate based on `InstallProtocolInterface()`.

The `r-efi` crate is really only an attempt to provide the raw UEFI definitions and protocols, hence such an helper would be misplaced here. However, we already have several other helper crates like `r-efi-alloc` or `r-efi-string`, and I think something like `r-efi-util` is definitely worth it. However, what I would rather like to see is a proper modeling of the UEFI objects as real rust structs and traits, providing a proper application interface.

Long story short: I would love to see rust core support variadics for anything other than `extern "C"`. Without that, though, I don't see what we can do here other than document the alternatives. Do you have other suggestions what to do here?

Btw., I am on an hiatus until 2020, so I will not be able to work on the rust core fixes before that. Sorry!

**Comment by jyao1** [comment]
Understood. That makes sense.
It is hard to support variadics without compiler support.
I will close it.
